sys/senml: add SenML modules

silkeh opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

Add a basic SenML module with support for:
    
- Encoding SenML values as CBOR using NanoCBOR.
- Converting from Phydat to SenML.
- Reading and encoding SAUL sensors.

Some effort is made to convert units from those in Phydat to those supported by SenML, which results in some code that should probably not be part of a SenML implementation (including calculation of Unix timestamps).

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

Includes two test cases and an example. The `senml_saul` test isn't useful at this point, but I couldn't find if there are some mocked sensors already.

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

Depends on https://github.com/bergzand/NanoCBOR/pull/50

kfessel commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,179 @@
+{
+    senml->attr.unit = SENML_UNIT_SECOND;
+
+    int64_t month = (int64_t)phydat->val[0];
+    int64_t year = (int64_t)phydat->val[0];
+
+    /* Calculate days of the year based on the month and leap year */
+    int64_t day = (int64_t)phydat->val[0] + month_to_yday[month] +
```
something seems wrong here all information is in  phydat->val[0] ??

silkeh replied:
There were some more errors in this part of the code. I have fixed it, and added a few test cases for `senml_phydat` including ones for unix time calculation.

fjmolinas commented on sys/include/senml.h:
```diff
@@ -0,0 +1,377 @@
+    senml_unit_t base_unit;         /**< Base Unit */
+    senml_numeric_t base_value;     /**< Base Value */
+    senml_numeric_t base_sum;       /**< Base Sum */
+    uint64_t base_version;          /**< Base Version */
```
Does it make sense to have compile-time configuration remove some of the optional fields and save some RAM?, specifically `base_version`, `update_time`, `sum`, which will be used more rarely.

fjmolinas commented on sys/include/senml.h:
```diff
@@ -0,0 +1,377 @@
+ * @param v Value to encode.
+ * @return Numeric value containing the given value.
+ */
+inline senml_numeric_t senml_float(float v)
```
```suggestion
static inline senml_numeric_t senml_float(float v)
```

And for other cases below.

fjmolinas replied:
Could we add static initializers as well? Some of the data structures are quite heavy (150bytes) and could be statically allocated to save stack space in many cases.

fjmolinas commented on sys/senml/senml.c:
```diff
@@ -0,0 +1,127 @@
+
+    case SENML_UNIT_BYTE:                         return "B";
+    case SENML_UNIT_VOLT_AMPERE:                  return "VA";
+    case SENML_UNIT_VOLT_AMPERE_SECOND:           return "VAs";
```
Can you add this false positive?

fjmolinas reviewed (commented):
In general, this looks nice. I'm not sure about the `senml_saul` module though. I'm not sure many users will need it since it reads all `saul_reg_t` by default, it seems like something that would be better fitted to be in the examples.

I think what would make more sense is having a:

```c
void senml_saulreg_encode(void *ctx, saul_reg_t *dev);
ssize_t senml_saulregs_encode(coid *ctx, saul_reg_t **dev, uint8_t num);
```

Where the second would be an array of `saul_regs_t`. I think in most cases the user would know how many `sensors` it will encode and therefore a definite array is a better solution, a convenience wrapper could be provided as well, so something like:

```c
ssize_t senml_saulregs_encode_cbor(uint8_t *buf, size_t len, saul_reg_t **dev, uint8_t num)
{
    nanocbor_encoder_t enc;

    nanocbor_encoder_init(&enc, buf, len);
    nanocbor_fmt_array(&enc, num);
    for (uint8_t i = 0; i < num; i++) {
        senml_saulreg_encode(&enc, *dev++);
    }
    return nanocbor_encoded_len(&enc);
}
```

bergzand commented on pkg/nanocbor/Makefile:
```diff
@@ -1,6 +1,6 @@
 PKG_NAME    = nanocbor
 PKG_URL     = https://github.com/bergzand/nanocbor
-PKG_VERSION = 97e8cbe943388dda0087ab703034fec268c7c12d
+PKG_VERSION = 57c48edaca2b5792a41119610035335761dd3c81
```
This should not be necessary anymore with #16538 merged

silkeh replied:
I will drop that commit when squashing/rebasing

fjmolinas commented:
This is no longer waiting. Can you rebase @silkeh

Citrullin commented on sys/include/senml/saul.h:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @return Size of the encoded data.
+ */
+size_t senml_saul_encode_cbor(uint8_t *buf, size_t len);
```
I would find it very helpful to have an API for individual devices as well. As far as I am aware of, you can also have specific endpoints for specific sensors. Like /temp, /humidity etc. 
So, in that case you want only to encode the temperature sensor etc. for the /temp endpoint.

Citrullin commented:
> 
> 
> In general, this looks nice. I'm not sure about the `senml_saul` module though. I'm not sure many users will need it since it reads all `saul_reg_t` by default, it seems like something that would be better fitted to be in the examples.
> 
> I think what would make more sense is having a:
> 
> ```c
> void senml_saulreg_encode(void *ctx, saul_reg_t *dev);
> ssize_t senml_saulregs_encode(coid *ctx, saul_reg_t **dev, uint8_t num);
> ```
> 
> Where the second would be an array of `saul_regs_t`. I think in most cases the user would know how many `sensors` it will encode and therefore a definite array is a better solution, a convenience wrapper could be provided as well, so something like:
> 
> ```c
> ssize_t senml_saulregs_encode_cbor(uint8_t *buf, size_t len, saul_reg_t **dev, uint8_t num)
> {
>     nanocbor_encoder_t enc;
> 
>     nanocbor_encoder_init(&enc, buf, len);
>     nanocbor_fmt_array(&enc, num);
>     for (uint8_t i = 0; i < num; i++) {
>         senml_saulreg_encode(&enc, *dev++);
>     }
>     return nanocbor_encoded_len(&enc);
> }
> ```

I agree. I would also like to see an API for individual devices. This would be very helpful.

silkeh commented:
> In general, this looks nice. I'm not sure about the `senml_saul` module though. I'm not sure many users will need it since it reads all `saul_reg_t` by default, it seems like something that would be better fitted to be in the examples.
> 
> I think what would make more sense is having a:
> 
> ```c
> void senml_saulreg_encode(void *ctx, saul_reg_t *dev);
> ssize_t senml_saulregs_encode(coid *ctx, saul_reg_t **dev, uint8_t num);
> ```
> 
> Where the second would be an array of `saul_regs_t`. I think in most cases the user would know how many `sensors` it will encode and therefore a definite array is a better solution, a convenience wrapper could be provided as well, so something like:
> ```c
> ssize_t senml_saulregs_encode_cbor(uint8_t *buf, size_t len, saul_reg_t **dev, uint8_t num)
> {
>     nanocbor_encoder_t enc;
> 
>     nanocbor_encoder_init(&enc, buf, len);
>     nanocbor_fmt_array(&enc, num);
>     for (uint8_t i = 0; i < num; i++) {
>         senml_saulreg_encode(&enc, *dev++);
>     }
>     return nanocbor_encoded_len(&enc);
> }
> ```

SAUL is structured in such a way that the dimensions of sensor values are entirely unknowable until you actually read it. I think it can even change on every read (theoretically). This means that the number of sensors and the number of SenML values have a ratio of somewhere between `1` and `PHYDAT_DIM`, and we cannot determine beforehand how many SenML values will be returned.

As for passing a list with a number of sensors: is it common to have a list in such a way? Why not just pass a `saul_reg_t *dev`, as that's a linked list already?

Taking the above into account, I have split the implementation in two functions:

```c
int senml_saul_reg_encode_cbor(nanocbor_encoder_t *enc, saul_reg_t *dev);
size_t senml_saul_encode_cbor(uint8_t *buf, size_t len, saul_reg_t *reg);
```

The first function encodes all dimensions of a SAUL sensor (and returns the number of dimensions encoded), and the second one is encodes all functions in a registry. The second one is extremely simple now, and I think it would be fine to move that to the example if so desired.

Encoding all the temperature values would then look something like this:
```c
nanocbor_encoder_t enc;

nanocbor_encoder_init(&enc, buf, len);
nanocbor_fmt_array_indefinite(&enc);

saul_reg_t *dev = saul_reg;
while (dev) {
    if(dev->driver->type = SAUL_SENSE_TEMP) {
        senml_saul_reg_encode_cbor(&enc, dev);
    }
    dev = dev->next;
}

nanocbor_fmt_end_indefinite(&enc);
```

Citrullin commented:
> SAUL is structured in such a way that the dimensions of sensor values are entirely unknowable until you actually read it. I think it can even change on every read (theoretically). This means that the number of sensors and the number of SenML values have a ratio of somewhere between `1` and `PHYDAT_DIM`, and we cannot determine beforehand how many SenML values will be returned.
> 
> As for passing a list with a number of sensors: is it common to have a list in such a way? Why not just pass a `saul_reg_t *dev`, as that's a linked list already?

I think a point you forget is that a user can also create a custom linked list of `saul_reg_t`. Additionally to the default SAUL register. This may be even needed for certain use-cases. Or, like in my case, I need to individually encode each sensor.
I don't want to have one endpoint with all sensors. I need them under seperated endpoints. With the current design of your API, I can't do my use case. I don't mind having a different API for it. I just need to be able to encode each device individually and a custom list.

> Taking the above into account, I have split the implementation in two functions:
> 
> ```c
> int senml_saul_reg_encode_cbor(nanocbor_encoder_t *enc, saul_reg_t *dev);
> size_t senml_saul_encode_cbor(uint8_t *buf, size_t len, saul_reg_t *reg);
> ```
> 
> The first function encodes all dimensions of a SAUL sensor (and returns the number of dimensions encoded), and the second one is encodes all functions in a registry. The second one is extremely simple now, and I think it would be fine to move that to the example if so desired.
> 

For the second function: Why not having `nanocbor_encoder_t ` here as well. I am not familiar with nanocbor. Can't you reuse the encoder instead of initializing another one for each device you may encode?

silkeh commented:
> > SAUL is structured in such a way that the dimensions of sensor values are entirely unknowable until you actually read it. I think it can even change on every read (theoretically). This means that the number of sensors and the number of SenML values have a ratio of somewhere between `1` and `PHYDAT_DIM`, and we cannot determine beforehand how many SenML values will be returned.
> > As for passing a list with a number of sensors: is it common to have a list in such a way? Why not just pass a `saul_reg_t *dev`, as that's a linked list already?
> 
> I think a point you forget is that a user can also create a custom linked list of `saul_reg_t`. Additionally to the default SAUL register. This may be even needed for certain use-cases. Or, like in my case, I need to individually encode each sensor.
> I don't want to have one endpoint with all sensors. I need them under seperated endpoints.

So why not create a custom registry instead of a linked list? The current implementation allows you to pass any `saul_reg_t`, including custom ones. For actually custom data structures I think it's fine to have to to a bit more effort. Especially because (depending on the use case), you can optimize a bit: for example, the following would encode a list of `saul_reg_t` sensors (with known dimensions):

```c
nanocbor_encoder_t enc;
nanocbor_encoder_init(&enc, buf, len);
nanocbor_fmt_array(&enc, sensor_count);
for (int i = 0; i < sensor_count; i++) {
    senml_saul_reg_encode_cbor(&enc, dev);
}

return nanocbor_encoded_len(&enc)
```

> > Taking the above into account, I have split the implementation in two functions:
> > ```c
> > int senml_saul_reg_encode_cbor(nanocbor_encoder_t *enc, saul_reg_t *dev);
> > size_t senml_saul_encode_cbor(uint8_t *buf, size_t len, saul_reg_t *reg);
> > ```
> > 
> >   
> > The first function encodes all dimensions of a SAUL sensor (and returns the number of dimensions encoded), and the second one is encodes all functions in a registry. The second one is extremely simple now, and I think it would be fine to move that to the example if so desired.
> 
> For the second function: Why not having `nanocbor_encoder_t ` here as well. I am not familiar with nanocbor. Can't you reuse the encoder instead of initializing another one for each device you may encode?

Sure, but just use something like the example above. The code is sufficiently composable to do whatever you like.

fjmolinas commented:
Sorry for dropping the ball on this one I'll try to jump back on this early next week, would be nice to get it in for this Release.

kfessel commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,185 @@
+    int64_t year = (int64_t)phydat->val[2] + (month > 1) - 1900;
+
+    /* Calculate days of the year based on the month */
+    int64_t day = (int64_t)phydat->val[0] + (int64_t)month_to_yday[month] - 1;
```
64bit of month, year and day seems a bit much

fjmolinas replied:
seems excessive indeed...

fjmolinas commented on sys/include/senml/saul.h:
```diff
@@ -0,0 +1,76 @@
+ * @ref senml_numeric_t using @ref phydat_to_senml_float.
+ * Values are converted using @ref phydat_to_senml_decimal otherwise.
+ */
+#ifndef SENML_SAUL_USE_FLOATS
+#define SENML_SAUL_USE_FLOATS (0)
+#endif
```
Since this is a configuration can you prefix them with `CONFIG_`?
```suggestion
#ifndef CONFIG_SENML_SAUL_USE_FLOATS
#define CONFIG_SENML_SAUL_USE_FLOATS (0)
#endif
```

fjmolinas commented on sys/include/senml.h:
````diff
@@ -0,0 +1,501 @@
+ * see @ref senml_attr_t for details. To enable all attributes, set:
+ *
+ * ```
+ * CFLAGS += -DSENML_ATTR_SUM
+ * CFLAGS += -DSENML_ATTR_VERSION
+ * CFLAGS += -DSENML_ATTR_UPDATE_TIME
````
Can you add `ifndef/define` for these and prefix them with `CONFIG_`?

silkeh replied:
I have prefixed them with `CONFIG_`, but I don't understand what you want with an `ifndef/define`?

fjmolinas replied:
I mean that you never actually defines these, you rely on doing nothing if they are not defined, which works, but does not allow me to search for them in doxygen. So I would like to have something like

```
#ifndef CONFIG_SENML_ATTR_SUM
#define CONFIG_SENML_ATTR_SUM   0
#ifndef
```

(Thar would require switching all the `ifdef  CONFIG_SENML_ATTR_SUM` to `if IS_ACTIVE(CONFIG_SENML_ATTR_SUM)`)

Or otherwise something like:

```
/**
 * @brief   Define SenML update time attribute, disabled by default
 */
#if defined(DOXYGEN)
#define CONFIG_SENML_ATTR_UPDATE_TIME
#endif
```

fjmolinas commented on sys/include/senml.h:
```diff
@@ -0,0 +1,501 @@
+ */
+typedef struct {
+    senml_value_type_t type; /**< Type of the value */
+    union {
+        uint64_t u;
+        int64_t i;
+        float f;
+        double d;
+        struct { int32_t e; int32_t m; } df /** Decimal fraction */;
```
I think Doxygen will complain here once you rebase...

silkeh replied:
So?

fjmolinas replied:
You will need to comment all members of the union. But rebase and see I guess.

silkeh replied:
No complaints :shrug:

fjmolinas commented on sys/include/senml.h:
```diff
@@ -0,0 +1,501 @@
+ */
+typedef struct {
+    senml_value_type_t type; /**< Type of the value */
+    union {
+        uint64_t u;
+        int64_t i;
+        float f;
+        double d;
+        struct { int32_t e; int32_t m; } df /** Decimal fraction */;
```
Can you unscrustify?

silkeh replied:
This is uncrustified.

fjmolinas replied:
```suggestion
    union {
        uint64_t u;
        int64_t i;
        float f;
        double d;
        struct {
            int33_t e;
            int32_t m;
        } df /** Decimal fraction */;
```

I guess I was expecting it to change to the above, but if it does not mind leave it as is.

fjmolinas commented on sys/include/senml.h:
```diff
@@ -0,0 +1,501 @@
+    senml_numeric_t base_time;      /**< Base Time */
+    senml_unit_t base_unit;         /**< Base Unit */
+    senml_numeric_t base_value;     /**< Base Value */
+#if defined(SENML_ATTR_SUM) || defined(DOXYGEN)
```
See comment above regarding `CONFIG_`

fjmolinas commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,185 @@
+
+    /* simple conversions */
+    case UNIT_TEMP_F:
+        value = (value + 459.67) * (5. / 9.);
```
Magic numbers?

silkeh replied:
I have added a comment

fjmolinas commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,185 @@
+        senml->attr.unit = SENML_UNIT_KELVIN;
+        break;
+    case UNIT_G:
+        value *= 9.80665;
```
Magic numbers?

silkeh replied:
I have added a comment

fjmolinas reviewed (commented):
Some final nitpicks. Otherwise this seems good to me, I'll try to provide a patch for Kconfig.

fjmolinas commented on sys/include/senml/phydat.h:
```diff
@@ -0,0 +1,120 @@
+ * @ingroup     sys_senml
+ * @brief       Functionality for converting from @ref sys_phydat to @ref sys_senml
+ *
+ * The `senml_phydat` module contains various functions for converting @ref sys_phydat values to @ref sys_senml.
```
Mind breaking this line?

fjmolinas commented:
This Kconfig patch should do the trick:

```
diff --git a/sys/Kconfig b/sys/Kconfig
index 30e9feeb62..ffd6058f31 100644
--- a/sys/Kconfig
+++ b/sys/Kconfig
@@ -72,6 +72,7 @@ rsource "random/Kconfig"
 rsource "saul_reg/Kconfig"
 rsource "schedstatistics/Kconfig"
 rsource "sema/Kconfig"
+rsource "senml/Kconfig"
 rsource "seq/Kconfig"
 rsource "shell/Kconfig"
 rsource "test_utils/Kconfig"
diff --git a/sys/senml/Kconfig b/sys/senml/Kconfig
index 38a4189c1d..5c2a0a1d9e 100644
--- a/sys/senml/Kconfig
+++ b/sys/senml/Kconfig
@@ -8,6 +8,31 @@
 config MODULE_SENML
     bool "SenML"
     depends on TEST_KCONFIG
-    select MODULE_PHYDAT
     help
         Generic data container for physical data and utility functions.
+
+config MODULE_SENML_SAUL
+    bool "SenML SAUL  integration"
+    depends on TEST_KCONFIG
+    select MODULE_SENML_PHYDAT
+    select MODULE_SENML_CBOR
+    depends on MODULE_SAUL
+    depends on MODULE_SAUL_REG
+    help
+        Generic data container for physical data and utility functions.
+
+config MODULE_SENML_PHYDAT
+    bool "SenML Phydat support"
+    depends on TEST_KCONFIG
+    select MODULE_SENML
+    select MODULE_PHYDAT
+    help
+        Utilities to convert Phydat valus to SenML
+
+config MODULE_SENML_CBOR
+    bool "SenML CBOR enconding"
+    depends on TEST_KCONFIG
+    select MODULE_SENML
+    select PACKAGE_NANOCBOR
+    help
+        Support for CBOR encoding of SenML values
diff --git a/tests/senml_cbor/app.config.test b/tests/senml_cbor/app.config.test
index 1a938107b1..ff0dae5786 100644
--- a/tests/senml_cbor/app.config.test
+++ b/tests/senml_cbor/app.config.test
@@ -1 +1,3 @@
-CONFIG_MODULE_SENML_SAUL=y
+CONFIG_MODULE_SENML_CBOR=y
+CONFIG_MODULE_FMT=y
+CONFIG_MODULE_EMBUNIT=y
diff --git a/tests/senml_phydat/app.config.test b/tests/senml_phydat/app.config.test
new file mode 100644
index 0000000000..56d8e9a97c
--- /dev/null
+++ b/tests/senml_phydat/app.config.test
@@ -0,0 +1,2 @@
+CONFIG_MODULE_SENML_PHYDAT=y
+CONFIG_MODULE_EMBUNIT=y
diff --git a/tests/senml_saul/app.config.test b/tests/senml_saul/app.config.test
new file mode 100644
index 0000000000..5bccc9280b
--- /dev/null
+++ b/tests/senml_saul/app.config.test
@@ -0,0 +1,6 @@
+CONFIG_MODULE_SAUL_DEFAULT=y
+CONFIG_MODULE_SAUL=y
+CONFIG_MODULE_SAUL_REG=y
+CONFIG_MODULE_SENML_SAUL=y
+CONFIG_MODULE_EMBUNkIT=y
+CONFIG_MODULE_FMT=y
```

fjmolinas reviewed (commented):
On a third look,  can we move all the `tests/senml*` to `unittests? I think they should be all different tests suites of the same application. Kind of like for tests/unittests/tests-ztimer/tests-ztimer.c. What do you think?

silkeh commented:
> On a third look, can we move all the tests/senml* to `unittests? I think they should be all different tests suites of the same application. Kind of like for tests/unittests/tests-ztimer/tests-ztimer.c. What do you think?

I have no idea what you mean here. What is the difference between `tests` and `tests/unittests`? Shouldn't they be test suites of the respective modules instead of test suites of an application?

kfessel commented on sys/senml/phydat.c:
```diff
@@ -74,21 +76,22 @@ void phydat_date_to_senml(senml_value_t *senml, const phydat_t *phydat)
     /* Get year relative to 1900. */
     /* Add a year for months after Feb to deal with leap years. */
-    int64_t year = (int64_t)phydat->val[2] + (month > 1) - 1900;
+    int16_t year = (int64_t)(phydat->val[2] + (month > 1) - 1900);
```
is this a test for the compiler?

silkeh replied:
What do you mean?

kfessel commented on sys/senml/phydat.c:
```diff
@@ -74,21 +76,22 @@ void phydat_date_to_senml(senml_value_t *senml, const phydat_t *phydat)
 
     /* Calculate days of the year based on the month */
-    int64_t day = (int64_t)phydat->val[0] + (int64_t)month_to_yday[month] - 1;
+    int16_t day = (int64_t)phydat->val[0] + (int64_t)month_to_yday[month] - 1;
```
is this a test for the compiler?

silkeh replied:
What do you mean?

kfessel reviewed (commented):
something strange is happening with the types in the D-M-Y -> UNIX seconds conversion

Cast up and down 

Seems like some test for the compiler / optimizer and mem/code-sanatizer.

kfessel commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,190 @@
+
+    /* Get year relative to 1900. */
+    /* Add a year for months after Feb to deal with leap years. */
+    int16_t year = (int64_t)(phydat->val[2] + (month > 1) - 1900);
```
```suggestion
    int16_t year = phydat->val[2] + ((month > 1?)1:0) - 1900;
```

kfessel commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,190 @@
+    int16_t year = (int64_t)(phydat->val[2] + (month > 1) - 1900);
+
+    /* Calculate days of the year based on the month */
+    int16_t day = (int64_t)phydat->val[0] + (int64_t)month_to_yday[month] - 1;
```
```suggestion
    int16_t day = phydat->val[0] + month_to_yday[month] - 1;
```

kfessel commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,190 @@
+
+    /* POSIX calculation of a UNIX timestamp. */
+    /* See section 4.16 of the The Open Group Base Specifications Issue 7. */
+    senml->value = senml_int((int64_t)day * 86400 +
+                             (int64_t)(year - 70) * 31536000 +
+                             (int64_t)((year - 69) / 4) * 86400 -
+                             (int64_t)((year - 1) / 100) * 86400 +
+                             (int64_t)((year + 299) / 400) * 86400);
```
```suggestion
    int16_t leapdays = (((year - 69) / 4) - ((year - 1) / 100) + ((year + 299) / 400));
    senml->value = senml_int((int64_t)(day + leapdays)  * 86400 +
                             (int64_t)(year - 70) * 31536000;
```

fjmolinas commented on sys/senml/saul.c:
```diff
@@ -0,0 +1,129 @@
+
+static inline uint8_t senml_fix_unit(const saul_reg_t *dev, const uint8_t unit)
+{
+    // Fix the unit for relative humidity.
```
```suggestion
    /* Fix the unit for relative humidity. */
```

fjmolinas commented on tests/senml_cbor/Makefile:
```diff
@@ -0,0 +1,9 @@
+USEMODULE += fmt
+USEMODULE += embunit
+
+CFLAGS += -DSENML_ATTR_SUM -DSENML_ATTR_VERSION -DSENML_ATTR_UPDATE_TIME
```
```suggestion
CFLAGS += -DCONFIG_SENML_ATTR_SUM
CFLAGS += -DCONFIG_SENML_ATTR_VERSION
CFLAGS += -DCONFIG_SENML_ATTR_UPDATE_TIME
```

fjmolinas reviewed (commented):
> > On a third look, can we move all the tests/senml* to `unittests? I think they should be all different tests suites of the same application. Kind of like for tests/unittests/tests-ztimer/tests-ztimer.c. What do you think?
> 
> I have no idea what you mean here. What is the difference between `tests` and `tests/unittests`? Shouldn't they be test suites of the respective modules instead of test suites of an application?

Maybe there is no need to move them to unittests, it might bloat the files to much but what I meant was having the following:

- tests/senml/main.c

```c
int main(void)
{
    TESTS_START();
    TESTS_RUN(tests_senml_cbor());
    TESTS_RUN(tests_senml_phydat());
    TESTS_RUN(tests_senml_saul());
    TESTS_END();
    return 0;
}
```

- tests/senml_*

```c
static void test_senml_encode(void)
{
    size_t len = senml_saul_encode_cbor(cbor_buf, sizeof cbor_buf, saul_reg);

    /* Encode as hex */
    fmt_bytes_hex(result, cbor_buf, len);

    /* Compare hex result */
    TEST_ASSERT_EQUAL_INT(2 * len, sizeof expect - 1);
    TEST_ASSERT_EQUAL_INT(0, strncmp(expect, result, len));
}

Test *tests_senml_saul(void)
{
    EMB_UNIT_TESTFIXTURES(fixtures) {
        new_TestFixture(test_senml_encode),
    };
    EMB_UNIT_TESTCALLER(senml_tests, NULL, NULL, fixtures);
    return (Test *)&senml_tests;
}

```

<details><summary>patch to have the above</summary>


```diff
diff --git a/tests/senml/Makefile b/tests/senml/Makefile
new file mode 100644
index 0000000000..4a7c292f71
--- /dev/null
+++ b/tests/senml/Makefile
@@ -0,0 +1,14 @@
+include ../Makefile.tests_common
+
+USEMODULE += saul_default
+USEMODULE += senml_saul
+USEMODULE += senml_phydat
+USEMODULE += senml_cbor
+USEMODULE += fmt
+USEMODULE += embunit
+
+CFLAGS += -DCONFIG_SENML_ATTR_SUM
+CFLAGS += -DCONFIG_SENML_ATTR_VERSION
+CFLAGS += -DCONFIG_SENML_ATTR_UPDATE_TIME
+
+include $(RIOTBASE)/Makefile.include
diff --git a/tests/senml/main.c b/tests/senml/main.c
new file mode 100644
index 0000000000..0dd4972828
--- /dev/null
+++ b/tests/senml/main.c
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2021 Silke Hofstra
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       SenML test
+ *
+ * @author      Silke Hofstra <silke@slxh.eu>
+ *
+ * @}
+ */
+
+#include "embUnit.h"
+
+Test *tests_senml_cbor(void);
+Test *tests_senml_phydat(void);
+Test *tests_senml_saul(void);
+
+int main(void)
+{
+    TESTS_START();
+    TESTS_RUN(tests_senml_cbor());
+    TESTS_RUN(tests_senml_phydat());
+    TESTS_RUN(tests_senml_saul());
+    TESTS_END();
+    return 0;
+}
diff --git a/tests/senml_cbor/tests/01-run.py b/tests/senml/tests/01-run.py
similarity index 100%
rename from tests/senml_cbor/tests/01-run.py
rename to tests/senml/tests/01-run.py
diff --git a/tests/senml_cbor/main.c b/tests/senml/tests_senml_cbor.c
similarity index 87%
rename from tests/senml_cbor/main.c
rename to tests/senml/tests_senml_cbor.c
index f6aa22b2a6..d80f5ed4d1 100644
--- a/tests/senml_cbor/main.c
+++ b/tests/senml/tests_senml_cbor.c
@@ -27,13 +27,14 @@
 
 #define BUF_SIZE (128)
 
-uint8_t cbor_buf[BUF_SIZE];
-char result[2 * BUF_SIZE];
-char expect[] = "89A5216943424F522074657374221A608404D001616D07187802F953B0A206"
-                "0102F953B0A2060202183DA2060302183DA2060402C48220190267A2060504"
-                "F5A20606036752494F54204F53A20607084400010203A201626B6705183D";
-
-void test_senml_encode(void)
+static uint8_t cbor_buf[BUF_SIZE];
+static char result[2 * BUF_SIZE];
+static char expect[] =
+    "89A5216943424F522074657374221A608404D001616D07187802F953B0A206"
+    "0102F953B0A2060202183DA2060302183DA2060402C48220190267A2060504"
+    "F5A20606036752494F54204F53A20607084400010203A201626B6705183D";
+
+static void test_senml_encode(void)
 {
     nanocbor_encoder_t enc;
 
@@ -108,7 +109,7 @@ void test_senml_encode(void)
     TEST_ASSERT_EQUAL_INT(0, strncmp(expect, result, len));
 }
 
-Test *tests_senml(void)
+Test *tests_senml_cbor(void)
 {
     EMB_UNIT_TESTFIXTURES(fixtures) {
         new_TestFixture(test_senml_encode),
@@ -116,11 +117,3 @@ Test *tests_senml(void)
     EMB_UNIT_TESTCALLER(senml_tests, NULL, NULL, fixtures);
     return (Test *)&senml_tests;
 }
-
-int main(void)
-{
-    TESTS_START();
-    TESTS_RUN(tests_senml());
-    TESTS_END();
-    return 0;
-}
diff --git a/tests/senml_phydat/main.c b/tests/senml/tests_senml_phydat.c
similarity index 98%
rename from tests/senml_phydat/main.c
rename to tests/senml/tests_senml_phydat.c
index a970622c10..100d7ac2f0 100644
--- a/tests/senml_phydat/main.c
+++ b/tests/senml/tests_senml_phydat.c
@@ -268,8 +268,7 @@ void test_phydat_to_senml_decimal(void)
     }
 }
 
-
-Test *tests_senml(void)
+Test *tests_senml_phydat(void)
 {
     EMB_UNIT_TESTFIXTURES(fixtures) {
         new_TestFixture(test_phydat_date_to_senml),
@@ -280,11 +279,3 @@ Test *tests_senml(void)
     EMB_UNIT_TESTCALLER(senml_tests, NULL, NULL, fixtures);
     return (Test *)&senml_tests;
 }
-
-int main(void)
-{
-    TESTS_START();
-    TESTS_RUN(tests_senml());
-    TESTS_END();
-    return 0;
-}
diff --git a/tests/senml_saul/main.c b/tests/senml/tests_senml_saul.c
similarity index 80%
rename from tests/senml_saul/main.c
rename to tests/senml/tests_senml_saul.c
index 0cef596e27..5151b1d105 100644
--- a/tests/senml_saul/main.c
+++ b/tests/senml/tests_senml_saul.c
@@ -27,11 +27,11 @@
 
 #define BUF_SIZE (128)
 
-uint8_t cbor_buf[BUF_SIZE];
-char result[2 * BUF_SIZE];
-char expect[] = "9FFF";
+static uint8_t cbor_buf[BUF_SIZE];
+static char result[2 * BUF_SIZE];
+static char expect[] = "9FFF";
 
-void test_senml_encode(void)
+static void test_senml_encode(void)
 {
     size_t len = senml_saul_encode_cbor(cbor_buf, sizeof cbor_buf, saul_reg);
 
@@ -43,7 +43,7 @@ void test_senml_encode(void)
     TEST_ASSERT_EQUAL_INT(0, strncmp(expect, result, len));
 }
 
-Test *tests_senml(void)
+Test *tests_senml_saul(void)
 {
     EMB_UNIT_TESTFIXTURES(fixtures) {
         new_TestFixture(test_senml_encode),
@@ -51,11 +51,3 @@ Test *tests_senml(void)
     EMB_UNIT_TESTCALLER(senml_tests, NULL, NULL, fixtures);
     return (Test *)&senml_tests;
 }
-
-int main(void)
-{
-    TESTS_START();
-    TESTS_RUN(tests_senml());
-    TESTS_END();
-    return 0;
-}
diff --git a/tests/senml_cbor/Makefile b/tests/senml_cbor/Makefile
deleted file mode 100644
index 1785023ba8..0000000000
--- a/tests/senml_cbor/Makefile
+++ /dev/null
@@ -1,9 +0,0 @@
-include ../Makefile.tests_common
-
-USEMODULE += senml_cbor
-USEMODULE += fmt
-USEMODULE += embunit
-
-CFLAGS += -DSENML_ATTR_SUM -DSENML_ATTR_VERSION -DSENML_ATTR_UPDATE_TIME
-
-include $(RIOTBASE)/Makefile.include
diff --git a/tests/senml_cbor/app.config.test b/tests/senml_cbor/app.config.test
deleted file mode 100644
index ff0dae5786..0000000000
--- a/tests/senml_cbor/app.config.test
+++ /dev/null
@@ -1,3 +0,0 @@
-CONFIG_MODULE_SENML_CBOR=y
-CONFIG_MODULE_FMT=y
-CONFIG_MODULE_EMBUNIT=y
diff --git a/tests/senml_phydat/Makefile b/tests/senml_phydat/Makefile
deleted file mode 100644
index 6c14df0faa..0000000000
--- a/tests/senml_phydat/Makefile
+++ /dev/null
@@ -1,6 +0,0 @@
-include ../Makefile.tests_common
-
-USEMODULE += senml_phydat
-USEMODULE += embunit
-
-include $(RIOTBASE)/Makefile.include
diff --git a/tests/senml_phydat/app.config.test b/tests/senml_phydat/app.config.test
deleted file mode 100644
index 56d8e9a97c..0000000000
--- a/tests/senml_phydat/app.config.test
+++ /dev/null
@@ -1,2 +0,0 @@
-CONFIG_MODULE_SENML_PHYDAT=y
-CONFIG_MODULE_EMBUNIT=y
diff --git a/tests/senml_phydat/tests/01-run.py b/tests/senml_phydat/tests/01-run.py
deleted file mode 100755
index 8b4e23f2e7..0000000000
--- a/tests/senml_phydat/tests/01-run.py
+++ /dev/null
@@ -1,14 +0,0 @@
-#!/usr/bin/env python3
-
-# Copyright (C) 2017 Freie Universität Berlin
-#
-# This file is subject to the terms and conditions of the GNU Lesser
-# General Public License v2.1. See the file LICENSE in the top level
-# directory for more details.
-
-import sys
-from testrunner import run_check_unittests
-
-
-if __name__ == "__main__":
-    sys.exit(run_check_unittests())
diff --git a/tests/senml_saul/Makefile b/tests/senml_saul/Makefile
deleted file mode 100644
index a69acec8c5..0000000000
--- a/tests/senml_saul/Makefile
+++ /dev/null
@@ -1,8 +0,0 @@
-include ../Makefile.tests_common
-
-USEMODULE += saul_default
-USEMODULE += senml_saul
-USEMODULE += fmt
-USEMODULE += embunit
-
-include $(RIOTBASE)/Makefile.include
diff --git a/tests/senml_saul/tests/01-run.py b/tests/senml_saul/tests/01-run.py
deleted file mode 100755
index 8b4e23f2e7..0000000000
--- a/tests/senml_saul/tests/01-run.py
+++ /dev/null
@@ -1,14 +0,0 @@
-#!/usr/bin/env python3
-
-# Copyright (C) 2017 Freie Universität Berlin
-#
-# This file is subject to the terms and conditions of the GNU Lesser
-# General Public License v2.1. See the file LICENSE in the top level
-# directory for more details.
-
-import sys
-from testrunner import run_check_unittests
-
-
-if __name__ == "__main__":
-    sys.exit(run_check_unittests())
```
</details>

But feel free to disregard it, I can propose this as a patch once this is in, feel free to squash.

fjmolinas commented:
> But feel free to disregard it, I can propose this as a patch once this is in, feel free to squash.

And rebase!

fjmolinas commented:
@silkeh BTW let me know if you are short on time and need some help, I can do the rebase and force push :)

silkeh commented:
> But feel free to disregard it, I can propose this as a patch once this is in, feel free to squash.

Thanks! I have squashed, rebased, fought with codespell, and I believe it should be OK now.

fjmolinas commented:
Forced push to add insufficient memory, the `PRIi64` issue seems to be only for AVR, the macro is undefined for that. I have multiple solutions but not sure whats the best, will discuss offline and comeback.

maribu commented:
The compilation failure is due to a bug in avrlibc. In `/usr/avr/include/inttypes.h` it says

```
#ifdef __avr_libc_does_not_implement_long_long_in_printf_or_scanf 
 
#define         PRId64                  "lld"
#define         PRIdLEAST64             "lld"
#define         PRIdFAST64              "lld"
 
#define         PRIi64                  "lli"
#define         PRIiLEAST64             "lli"
#define         PRIiFAST64              "lli"
 
 
#define         PRIdMAX                 "lld"
#define         PRIiMAX                 "lli"
 
#endif
```

It should say:

```
#ifndef __avr_libc_does_not_implement_long_long_in_printf_or_scanf 
 
#define         PRId64                  "lld"
#define         PRIdLEAST64             "lld"
#define         PRIdFAST64              "lld"
 
#define         PRIi64                  "lli"
#define         PRIiLEAST64             "lli"
#define         PRIiFAST64              "lli"
 
 
#define         PRIdMAX                 "lld"
#define         PRIiMAX                 "lli"
 
#endif
```

(`#ifndef` instead of `#ifdef`, not the double negation.)

The problem is that avrlibc is unmaintained and patches are not acted upon anymore.

maribu commented:
On Arduino Mega2560 with the above fix in `inttypes.h` of avrlibc:

```
2022-01-20 12:28:49,886 # Help: Press s to start test, r to print it is ready
s
2022-01-20 12:28:52,157 # START
2022-01-20 12:28:52,222 # main(): This is RIOT! (Version: 2022.04-devel-76-g35245-senml)
2022-01-20 12:28:52,227 # ...
2022-01-20 12:28:52,325 # senml_tests.test_phydat_to_senml_float (tests/senml_phydat/main.c 244) exp -10752 was -10753
2022-01-20 12:28:52,326 # .
2022-01-20 12:28:52,345 # run 4 failures 1
```

fjmolinas commented:
> The compilation failure is due to a bug in avrlibc. In /usr/avr/include/inttypes.h it says

Should we just add the define then?

maribu commented:
> Should we just add the define then?

I guess adding

```C
#ifndef PRIi64
#define PRIi64 "lli"
#endif
```

and similar should not hurt, even if upstream fixes the issue eventually. (I still hope someone revives avrlibc. It is a nifty platform for teaching - you hardly find hardware simpler and easier to explain than AVR.)

fjmolinas commented:
Thanks for testing @maribu I think I had only tested on 32bit, both 16bit and 8bit boards seems to have issues, something must be off.

maribu commented:
Patch submitted upstream: https://savannah.nongnu.org/patch/index.php?10171

fjmolinas commented:
> 

I think for the `phydat` part its mainly a matter of how rounding is done in the different toolchains...

fjmolinas commented on tests/senml_phydat/main.c:
```diff
@@ -0,0 +1,290 @@
+
+        phydat_to_senml_float(&res, &(value_tests[i].phydat), value_tests[i].dim);
+
+        DEBUG("Float: %" PRIi16 "e%" PRIi16 " %s -> %.f %s\n",
+              value_tests[i].phydat.val[value_tests[i].dim], value_tests[i].phydat.scale,
+              phydat_unit_to_str(value_tests[i].phydat.unit),
+              res.value.value.f,
+              senml_unit_to_str(res.attr.unit));
```
```suggestion
        if (IS_ACTIVE(ENABLE_DEBUG)) {
            print_str("Float: ");
            print_s32_dec(value_tests[i].phydat.val[value_tests[i].dim]);
            print_str("e");
            print_s32_dec(value_tests[i].phydat.scale);
            print_str(" ");
            print_str(phydat_unit_to_str(value_tests[i].phydat.unit));
            print_str(" -> ");
            print_float(res.value.value.f, 7);
            print_str(" ");
            print_str(phydat_unit_to_str(res.attr.unit));
            print_str("\n");
        }
```

Debugging I remembered non native can't print float, so I did this locally to investigate.

maribu replied:
Note that the pseudo module `printf_float` unlocks printing of floats. At least on AVR that works as advertised (or did so when I last tested).

fjmolinas commented:
The cbor tests just triggers a reset on both msp430 (z1) and avr (arduino-mega2560). I'll try to investigate a bit...

chrysn commented:
Looking at the API I perceived a large portion of the phydat handling to revolve about dates and times. Are they actually used anywhere, are they still untested, and given that SenML has no unit for point-in-time, only the seconds (without epoch relation) in values and the t values, does that conversion even make sense?

chrysn commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,189 @@
+    case UNIT_GR:      return SENML_UNIT_GRAM;
+    case UNIT_A:       return SENML_UNIT_AMPERE;
+    case UNIT_V:       return SENML_UNIT_VOLT;
+    case UNIT_W:       return SENML_UNIT_WATT;
```
```suggestion
    case UNIT_W:       return SENML_UNIT_WATT;
    case UNIT_T:       return SENML_UNIT_TESLA;
```

for when this is rebased after #17569

silkeh commented:
> Looking at the API I perceived a large portion of the phydat handling to revolve about dates and times. Are they actually used anywhere, are they still untested, and given that SenML has no unit for point-in-time, only the seconds (without epoch relation) in values and the t values, does that conversion even make sense?

Not sure if `UNIT_DATE` is actually produced by a sensor (perhaps an RTC?), but I think it makes sense to support as much of phydat as possible. The conversions have unit test coverage, so I'm fairly certain they work :slightly_smiling_face:.

As for using the Unix epoch: SenML explicitly uses seconds since the Unix epoch for absolute time (see RFC8428 [section 2](https://datatracker.ietf.org/doc/html/rfc8428#section-2) and [section 4.5.3](https://datatracker.ietf.org/doc/html/rfc8428#section-4.5.3)), so I think it makes sense to use that for dates as well.

chrysn commented:
> Not sure if `UNIT_DATE` is actually produced by a sensor (perhaps an RTC?)

It is not, and has never been used by anything inside the RIOT tree
except constructed tests.

> but I think it makes sense to support as much of phydat as possible.
> The conversions have unit test coverage, so I'm fairly certain they
> work :slightly_smiling_face:.

Good to know, I only remembered the "untested" line from the initial
post.

> As for using the Unix epoch: SenML explicitly uses seconds since the
> Unix epoch [...], so I think it makes sense to use that for dates as
> well. 

It uses it for `t`, but the UNIT_DATE values are converted into
senml_value_t and assigned a SenML unit, so they are interpreted as
values -- and nothing in the RFCs indicates such seconds would be rooted
here or there.

Overall, not objecting the conversions' presence here; I'm merely
cautioning that they make assumptions on convertability that I prefer
not to be done lightly, and suggesting that refusing to converting
things that have no well-founded justification for conversion might be a
feature of this new module at least as valuable as completeness of
conversion might be.

silkeh commented:
> > Not sure if `UNIT_DATE` is actually produced by a sensor (perhaps an RTC?)
>
> It is not, and has never been used by anything inside the RIOT tree except constructed tests.

In that case the date and time conversions could be removed entirely, as far as I'm concerned.
Although I think having date to Unix timestamp functionality would probably be useful to have in RIOT somewhere else (probably in phydat itself), in case one wants to set them from an RTC when using SenML (but that's a matter for a different PR).

> > but I think it makes sense to support as much of phydat as possible. The conversions have unit test coverage, so I'm fairly certain they work :slightly_smiling_face:.
>
> Good to know, I only remembered the "untested" line from the initial post.

Whoops, seems I forgot to remove that after I added tests.

> > As for using the Unix epoch: SenML explicitly uses seconds since the Unix epoch [...], so I think it makes sense to use that for dates as well.
>
> It uses it for `t`, but the UNIT_DATE values are converted into senml_value_t and assigned a SenML unit, so they are interpreted as values -- and nothing in the RFCs indicates such seconds would be rooted here or there.

Fair enough, that's an implementation decision made here keeping the usage in SenML in mind. 

> Overall, not objecting the conversions' presence here; I'm merely cautioning that they make assumptions on convertability that I prefer not to be done lightly, and suggesting that refusing to converting things that have no well-founded justification for conversion might be a feature of this new module at least as valuable as completeness of conversion might be.

Thinking about it a bit, I don't think the conversions make much sense anyway:

- Converting a date to a Unix timestamp brings along timezone issues, as it's currently assumed to be UTC.
- Converting date and time separately and having to add them afterwards is a bit convoluted.
- Converting a time to a number of seconds is sensible if it represents a duration, but nonsense when using a time of day (unless added to a Unix timestamp, which brings back the timezone issues).

How do you feel about removing the conversions from this PR, and creating a PR implementing Unix time functionality to phydat? I'm thinking about something like:

```c
int64_t phydate_date_time_to_unix(phydat_t *date, phydat_t *time, int32_t offset_seconds);
```

chrysn commented:
> How do you feel about removing the conversions from this PR,

+1

> ```c
> int64_t phydate_date_time_to_unix(phydat_t *date, phydat_t *time, int32_t offset_seconds);
> ```

That (or the split-up version) sounds like a good place to have that
code if you still want to provide it.

In the context of CBOR it may be convenient to tie that any of the terms
of <https://www.ietf.org/archive/id/draft-ietf-cbor-time-tag-00.html>
(which is currently not covering time zones, but that might change
depending on what happens in SEDATE around
<https://www.ietf.org/archive/id/draft-ietf-sedate-datetime-extended-02.html>).

chrysn commented on sys/senml/phydat.c:
```diff
@@ -0,0 +1,189 @@
+    case UNIT_TIME:    return SENML_UNIT_NONE;      /* split into second/minute/hour */
+    case UNIT_DATE:    return SENML_UNIT_NONE;      /* split into day/month/year */
+    case UNIT_GPM3:    return SENML_UNIT_NONE;      /* use kg/m3 instead */
+    case UNIT_DPS:     return SENML_UNIT_NONE;      /* no alternative */
+    case UNIT_CPM3:    return SENML_UNIT_NONE;      /* no alternative */
```
These two would make fine additions to the SenML registry (CPM3 as primary, and DPS as secondary to a to-be-added rad/s, see https://github.com/RIOT-OS/RIOT/issues/17571#issuecomment-1022161280). The registry is rather open, a mail to core@ietf.org has sufficed in the past. Would you care to just add them to SenML?

(Yes, this is a blunt attempt to get RIOT people more familiar with IETF workings ;-) )

silkeh replied:
I think having these added to the SenML registry is a great idea, but maybe we should do that as part of #17571 so we can get this PR merged.

chrysn replied:
Sure, doesn't need to be in this PR -- but anyway, the change would not happen until the appointed experts have approved the registry extension. Would you care to make the request?

silkeh replied:
Sure, I have added a proposal to #17571 to allow others to comment before I submit them though.

kaspar030 commented:
I'm testing `tests/senml_cbor` on z1. it reboots unless increasing stack size. I tried `CFLAGS += -DTHREAD_STACKSIZE_DEFAULT=1536`, that makes it pass.

fjmolinas commented:
> I'm testing `tests/senml_cbor` on z1. it reboots unless increasing stack size. I tried `CFLAGS += -DTHREAD_STACKSIZE_DEFAULT=1536`, that makes it pass.

Nice then its only a stack issue

fjmolinas commented:
> CFLAGS += -DTHREAD_STACKSIZE_DEFAULT=1536

Did the same thing or `arduino-mega2560`, test fails because encoded length is not as expected, form looking of the encoded payload the double is off:

```
2022-02-01 12:09:25,675 # START
2022-02-01 12:09:25,745 # main(): This is RIOT! (Version: 2022.04-devel-76-g35245-pr-16384)
2022-02-01 12:09:25,749 # .
2022-02-01 12:09:25,827 # senml_tests.test_senml_encode (tests/senml_cbor/main.c 107) exp 188 was 184
2022-02-01 12:09:25,827 # 
2022-02-01 12:09:25,847 # run 1 failures 1
```

```
[{-2: "CBOR test", -3: 1619264720, 1: "m", 7: 120, 2: 61.5}, {6: 1, 2: 2.46379138708083e-248}, 61, {6: 3, 2: 61}, {6: 4, 2: 4([-1, 615])}, {6: 5, 4: true}, {6: 6, 3: "RIOT OS"}, {6: 7, 8: h'00010203'}, {1: "kg", 5: 61}]
```

fjmolinas commented:
So this patch was enough to fix compilation issues with `avr`:

```diff
diff --git a/tests/senml_cbor/Makefile b/tests/senml_cbor/Makefile
index 971ce780ee..abac3c6b6a 100644
--- a/tests/senml_cbor/Makefile
+++ b/tests/senml_cbor/Makefile
@@ -8,4 +8,6 @@ CFLAGS += -DCONFIG_SENML_ATTR_SUM
 CFLAGS += -DCONFIG_SENML_ATTR_VERSION
 CFLAGS += -DCONFIG_SENML_ATTR_UPDATE_TIME
 
+CFLAGS += -DTHREAD_STACKSIZE_DEFAULT=1536
+
 include $(RIOTBASE)/Makefile.include
diff --git a/tests/senml_phydat/Makefile b/tests/senml_phydat/Makefile
index 6c14df0faa..452d330fda 100644
--- a/tests/senml_phydat/Makefile
+++ b/tests/senml_phydat/Makefile
@@ -2,5 +2,6 @@ include ../Makefile.tests_common
 
 USEMODULE += senml_phydat
 USEMODULE += embunit
+USEMODULE += printf_float
 
 include $(RIOTBASE)/Makefile.include
diff --git a/tests/senml_phydat/main.c b/tests/senml_phydat/main.c
index 6cc295d96f..76b38bad66 100644
--- a/tests/senml_phydat/main.c
+++ b/tests/senml_phydat/main.c
@@ -28,6 +28,10 @@
 #define ENABLE_DEBUG (0)
 #include "debug.h"
 
+#ifndef PRIi64
+#define PRIi64 "lli"
+#endif
+
 typedef struct {
     phydat_t phydat;
     senml_value_t senml;
```

But still both avr and msp430 are failing with rounding floats:

- avr
```
2022-02-01 12:21:51,830 # main(): This is RIOT! (Version: 2022.04-devel-76-g35245-pr-16384)
2022-02-01 12:21:51,834 # ...
2022-02-01 12:21:51,928 # senml_tests.test_phydat_to_senml_float (tests/senml_phydat/main.c 248) exp -10752 was -10753
2022-02-01 12:21:51,932 # .
2022-02-01 12:21:51,949 # run 4 failures 1
```

- msp430

```
2022-02-01 12:22:31,503 # START
2022-02-01 12:22:31,511 # main(): This is RIOT! (Version: 2022.04-devel-76-g35245-pr-16384)
2022-02-01 12:22:31,673 # ...
2022-02-01 12:22:31,685 # senml_tests.test_phydat_to_senml_float (tests/senml_phydat/main.c 247) exp 25537 was 0
2022-02-01 12:22:31,685 # .
2022-02-01 12:22:31,686 # run 4 failures 1
```

What test fails depends on what `round` function is used (at least for me locally), maybe we should skip that test case for some platforms.

fjmolinas commented:
> > CFLAGS += -DTHREAD_STACKSIZE_DEFAULT=1536
> 
> Did the same thing or `arduino-mega2560`, test fails because encoded length is not as expected, form looking of the encoded payload the double is off:
> 
> ```
> 2022-02-01 12:09:25,675 # START
> 2022-02-01 12:09:25,745 # main(): This is RIOT! (Version: 2022.04-devel-76-g35245-pr-16384)
> 2022-02-01 12:09:25,749 # .
> 2022-02-01 12:09:25,827 # senml_tests.test_senml_encode (tests/senml_cbor/main.c 107) exp 188 was 184
> 2022-02-01 12:09:25,827 # 
> 2022-02-01 12:09:25,847 # run 1 failures 1
> ```
> 
> ```
> [{-2: "CBOR test", -3: 1619264720, 1: "m", 7: 120, 2: 61.5}, {6: 1, 2: 2.46379138708083e-248}, 61, {6: 3, 2: 61}, {6: 4, 2: 4([-1, 615])}, {6: 5, 4: true}, {6: 6, 3: "RIOT OS"}, {6: 7, 8: h'00010203'}, {1: "kg", 5: 61}]
> ```

This could be a nanocbor issue with AVR since the `nanocbor_fmt_double` is not tested, maybe @bergzand can say more

bergzand commented on tests/senml_phydat/main.c:
```diff
@@ -0,0 +1,290 @@
+        .dim = 2
+    },
+    {
+        .phydat = { { 0, 0, 0 }, UNIT_TEMP_F, 3 },
+        .senml1 = senml_f(255.37, SENML_UNIT_KELVIN),
+        .senml2 = senml_df(0, 3, SENML_UNIT_NONE),
```
Quickly skimming over this, shouldn't these three match?
```suggestion
        .phydat = { { 25537, 0, 0 }, UNIT_TEMP_F, -2 },
        .senml1 = senml_f(255.37, SENML_UNIT_KELVIN),
        .senml2 = senml_df(25537, -2, SENML_UNIT_KELVIN),
```

fjmolinas replied:
Hmm `phydat` is in fahrenheit, `senml1` is in kelvin, so 0F is indeed 255.37K. For the `senml2` I think you are correct though since it would be converted to Kelvin, @silkeh is that correct?

fjmolinas replied:
I stand corrected, `senml2` is using `phydat_to_senml_decimal` which does not have special handling for `F` so it ends up calling ` phydat_unit_to_senml_unit` with no conversion.

bergzand replied:
Makes sense, sorry for the noise

silkeh commented:
> But still both avr and msp430 are failing with rounding floats:

Maybe we should only run the test on platforms with known predictable rounding:

```diff
diff --git a/tests/senml_phydat/main.c b/tests/senml_phydat/main.c
index 76b38bad66..02209fe596 100644
--- a/tests/senml_phydat/main.c
+++ b/tests/senml_phydat/main.c
@@ -278,7 +278,10 @@ Test *tests_senml(void)
     EMB_UNIT_TESTFIXTURES(fixtures) {
         new_TestFixture(test_phydat_date_to_senml),
         new_TestFixture(test_phydat_time_to_senml),
+/* Only run this test on CPUs with predictable rounding */
+#if defined(CPU_NATIVE) || defined(CPU_STM32)
         new_TestFixture(test_phydat_to_senml_float),
+# endif
         new_TestFixture(test_phydat_to_senml_decimal),
     };
     EMB_UNIT_TESTCALLER(senml_tests, NULL, NULL, fixtures);
```

If we really want to test if floats/rounding works predictably/reliably on all platforms there should probably be a different test for that.

fjmolinas commented:
> Maybe we should only run the test on platforms with known predictable rounding, eg:

I would rather go the other way around and not enable them for `MSP430` and `AVR`. But I agree this is not an issue with this PR. For me currently, the only thing I want to figure out is why `nanocbor_fmt_double` is failing on AVR, no need to fix it here but just figure out it's not an issue with this PRs code.

silkeh commented:
> I would rather go the other way around and not enable them for `MSP430` and `AVR`.

That's fine as well, although I'm having a bit of trouble finding what I should put in the `# if !defined()` statement to exclude `msp430` and `avr`

fjmolinas commented:
> > I would rather go the other way around and not enable them for `MSP430` and `AVR`.
> 
> That's fine as well, although I'm having a bit of trouble finding what I should put in the `# if !defined()` statement to exclude `msp430` and `avr`

For avr there is `#if defined(__AVR__)` for `msp430` there is `defined(CPU_MSP430FXYZ)`, not sure if there is another one, but those should do here,

silkeh commented:
I have pushed three fixup commits:

- Patch from @fjmolinas earlier.
- Removal of all date/time related stuff as requested by @chrysn.
- Exclusion of float tests for msp430 and AVR.

Please take a look and feel free to squash.

fjmolinas commented:
I pushed a couple of nitpicks

fjmolinas commented:
Rebased

fjmolinas commented:
Squashed

fjmolinas commented:
@chrysn Are you ok with getting this one in and addressing SenML/PhyDat unit issues separately?

fjmolinas reviewed (approved):
ACK, there are still some pending issues with:

- issues with handling of doubles on AVR, but this seems to come from either nanocbor or the CPU itself

- to limit builds jobs I would move all tests to a single application, this can be done in a follow-up, I have a branch with those somewhere (or a patch in the PR)

Other than this LGTM, I've tested on different architectures and BOARDs and except for the above mentioned issues and rounding differences on floats between platforms it all looks OK.

fjmolinas commented:
Force pushed missing Kconfig in test and some blacklisting.

fjmolinas commented:
Wuhuuu! Thanks for this one @silkeh and thanks for sticking through!
